### PR TITLE
Catalyst search filters

### PR DESCRIPTION
**Pull request by delphiactual**
Draft PR for catalyst filter.

 - [x] catalyst:complete
 - [x] catalyst:incomplete - missing objective information on socket
 - [x] catalyst:missing - missing enabled or disabled info on socket - all exotic weapons for some reason have a catalyst socket.

any and all assistance is appreciated.
This profile data should return the following:
| Incomplete | Missing |
|---|---|
| Outbreak Perfected | Dead Messenger |
| Whisper | MIDA |
| | Grand Overture |

all others are completed.

[profile-data.zip](https://github.com/DestinyItemManager/DIM/files/8193792/profile-data.zip)

### CODE REVIEW

**Comment by delphiactual** [comment]
`missing` is still out of picture until I can figure out how to check objective status on these

**Comment by delphiactual** [comment]
incomplete shows both missing and unslotted completed catalysts

**Comment by MeatSim** [comment]
Linking #8176, which also requires information from `Records`.

**Comment by bhollis** [comment]
I still don't quite understand what the search keywords should mean. Is this right?

* `catalyst:complete` - you have unlocked the catalyst, completed its objective, and applied the catalyst
* `catalyst:incomplete` - you have unlocked the catalyst, but have either not completed the objective or applied the catalyst
* `catalyst:missing` - the weapon has a potential catalyst, but you have not unlocked it

**Comment by bhollis** [comment]
OK, I think I have this in a good place. However, there's more to do and I'd like folks' opinions on what we should do:

1. There are "quest catalysts" which are different from catalysts you find in the world. These are generally seasonal exotics and there's a quest in your Pursuits for it (e.g. "Points Piercing Forever"). These catalysts aren't considered "unlocked" (in Triumphs) until you've completed that quest, unlike other catalysts which are unlocked as soon as you discover them, but aren't completed until you finish their objectives. We could do some pretty crazy stuff to fix that (basically, looking at quests in your inventory to see if you have the quest available) but maybe people generally understand how this works.
2. Given what we have here, we could remove catalyst sockets on exotic weapons where no catalyst exists.
3. We could start showing catalyst progress on the item popup - maybe like how we do with the Deepsight objective?

**Comment by andrewpong** [comment]
> OK, I think I have this in a good place. However, there's more to do and I'd like folks' opinions on what we should do:

I'm sure you guys have better ideas on how to do this without cluttering up the interface. I chucked a mock-up based on the Vendor inventory and here's my opinions FWIW:

1. A catalyst indicator on the item's icon is not useful and just clutters the Inventory view.
2. Status on the popup is useful, but doesn't need to be at the top.
3. It would be nice to have the catalyst' status on/near the Catalyst.
   - I think you'd just need to account for space for the catalyst and maybe an ornament, but then there's mementos, too... are there other edge cases...?
5. Some text to explain my catalyst progression would be nice so I put it at the bottom like the vendor collection text. 
   - I don't recall the name of that wierd section of the Triumphs page it's on - is catalyst progress considered a triumph, collectible, or something else? Is the backend and UI verbiage consistent from Bungie?
6. I like the succinct `catalyst:complete`, `catalyst:incomplete`, and `catalyst:missing`, as they convey enough to get people going in the right direction to acquire or complete their catalysts.
   - I recently confused myself and my blueberry friend when they asked me why some catalysts just dropped, while others required painful quests for non-pvpers like Witherhoard. I recall finally finishing Eriana's in S15 only to see the requirement to masterwork it. Oof.
7. Maybe there's value in `is:catalystexists`? 
   - Is there a use case to find exotic weapons which don't have a catalyst? 
   - The verb `has` is already used top check if something has a property (e.g. `haslight`, `hasmod`, `hasornament`), so that can't be used to check if something exists but is not collected. `exists` would be new, and adding new verbs shouldn't be done willy nilly, I expect.
   - I'm overthinking things, this has probably been solved for the similar `deepsight` filters.
9. Maybe the existing blue/green icons could work for this? 
   - No catalyst slot means there's no catalyst.
   - A catalyst slot without an icon: `catalyst:missing`.
   - A catalyst slot with a blue icon: `catalyst:incomplete`. 
   - A catalyst slot with a green icon: `catalyst:complete`.
10. Items 2-7 reminds me that DIM has good UX, I really admire what the DIM team has created, good UI/UX is _hard_, and that I should stick with my crayons and secops, hidden far from users.

image 

**Bonus request:** If it's maintainable, a little more information about how to acquire missing catalysts would be lovely, but this may be out of scope for DIM. 

Blueberry.gg has this info, but their article orientated website can take a lot more time to find info like this. Maybe a link to their page? Not sure if they have an API, they feel blog post-y. 

Examples:
* Rat King: "_This catalyst may be acquired by completing Vanguard Ops playlist strikes or Nightfalls. Kills in these activities have a small chance of dropping this catalyst.+" 
* Mida Multi-Tool, "_This catalyst may be acquired after attaining a Glory rank of Legend and completing a Competetive Crucible match._" 

Actually, the above highlights why I'd like to see catalyst acquisition info in DIM - there's too much conflicting and/or anecdotal information on how to get catalysts. I am not sure how accurate my Rat King acquisition blurb is, and is just a rewording of blueberries.gg: "Source: _Kills or completions in Strikes & Nightfalls._" 

I've been trying to help my blueberry friend acquire the Rat King catalyst but it just won't drop, but we don't do Nightfalls. A up to date and verified acquisition method for catalysts would be nice for catalysts not tied to special activities like Mida Multi-Tool, or catalysts quests as you described above (Witherhoard, Ticcu's, etc.).

But it depends on whether Catalyst acquisition info is within the scope of DIM, and who would be maintaining this information., 

Maybe catalyst details could be sourced in a manner similar to wishlists and their notes? Now this is sounding like a slippery slope towards an extension system for DIM, hah.

Relevant XKCD (just substitute "Standards" with "Catalyst Acquisition Procedure"):

image

**Comment by delphiactual** [comment]
Worldline Zero shows as incomplete...
image

image

image

[profile-data.zip](https://github.com/DestinyItemManager/DIM/files/9050264/profile-data.zip)

**Comment by delphiactual** [comment]
@andrewpong catalyst source info is available in-game on the catalyst pages, alas this information is not available via the API at this time.

image
image

**Comment by delphiactual** [comment]
> 1. There are "quest catalysts" which are different from catalysts you find in the world. These are generally seasonal exotics and there's a quest in your Pursuits for it (e.g. "Points Piercing Forever"). These catalysts aren't considered "unlocked" (in Triumphs) until you've completed that quest, unlike other catalysts which are unlocked as soon as you discover them, but aren't completed until you finish their objectives. We could do some pretty crazy stuff to fix that (basically, looking at quests in your inventory to see if you have the quest available) but maybe people generally understand how this works.
> 2. Given what we have here, we could remove catalyst sockets on exotic weapons where no catalyst exists.
> 3. We could start showing catalyst progress on the item popup - maybe like how we do with the Deepsight objective?

1. I think is understood behavior
2. Agree
3. Agree

**Comment by bhollis** [comment]
I'll have to check that Worldline catalyst - the data suggests that your profile is saying that record is incomplete - the check is very simple.

@andrewpong if you're looking to make a feature request please file it as such - this is a PR for a specific feature.

**Comment by bhollis** [comment]
OK, I fixed Worldline - weirdly the record said the objective wasn't complete AND the reward was unavailable, but the record was redeemed. I'm assuming this is because the catalyst isn't available in the game anymore. So I tweaked the logic to consider it complete as long as it's redeemed as well.

File: src/app/inventory/store/catalyst.ts [hunk 1]
```diff
@@ -27,5 +29,9 @@ export function buildCatalystInfo(
   // 2. could do a second pass on items populating it? rip through all the quest items, find ones whose rewards reference a catalyst, then go back and fix up the items' DimCatalyst info? would need a mapping from item hash to catalyst item hash (which I guess we can match up by name...)
   const unlocked = !(record.state & DestinyRecordState.Obscured);
 
+  if (itemHash === 1864563948) {
+    console.log(itemHash, { complete, unlocked, record });
+  }
+
```
Comment (delphiactual) [inline src/app/inventory/store/catalyst.ts:35]: remove?

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -226,6 +226,7 @@
 * Loadout Optimizer Mod picker will now correctly update when switching between mod slots without closing Mod Picker.
 * Loadout Optimizer now correctly takes Echo of Persistence's class-specific stat reductions into account when generating sets.
 * The "Kinetic Slot" icon in Compare sheet now looks different from the "Kinetic Damage" icon.
+* Added `catalyst:` filter which accepts the following parameters `all`, `complete`, and `incomplete`.
```
Comment (delphiactual) [inline docs/CHANGELOG.md:234]:
```suggestion
* Added `catalyst:` filter which accepts the following parameters `missing`, `complete`, and `incomplete`.
```

**Comment by delphiactual** [comment]
not sure why `jest test` is failing...

**Comment by robojumper** [comment]
> not sure why `jest test` is failing...

Probably because @nev-r pushed to master without updating the jest snapshot https://github.com/DestinyItemManager/DIM/commit/abb23e728a001d84f7b8a28c39fce421cbe39863

**Comment by nev-r** [comment]
ah yep, 2 for 1 bonus commit made it into the patch release fixes

**Comment by nev-r** [comment]
master has updated tests and this has no change collisions with that fix, but tests are still failing. might need merge?

**Comment by robojumper** [comment]
> but tests are still failing. might need merge?

I think a rerun of a workflow uses the merge base of the original run, so any commit that naturally triggers a new workflow should succeed test-wise.

**Review by robojumper — APPROVED** [review]
Seems to work fine in my testing, and in combination with #8604 this also works on the Records (Collections) page!

**Pull request merged**